src/Core/Services/StateService.cs

```diff
@@ -1363,6 +1363,15 @@ public void SetConfigs(ConfigResponse value)
             _storageMediatorService.Save(Constants.ConfigsKey, value);
         }
 
+        public async Task SetUserHasMasterPasswordAsync(bool value, string userId = null)
+        {
+            var reconciledOptions = ReconcileOptions(new StorageOptions { UserId = userId },
+                await GetDefaultStorageOptionsAsync());
+            var account = await GetAccountAsync(reconciledOptions);
+            account.Profile.UserDecryptionOptions.HasMasterPassword = value;
+            await SaveAccountAsync(account, reconciledOptions);
+        }
+
```

Comment:
🤔 Can this leave the app in a state where `HasMasterPassword` is `true` but for some error/cancellation of a user, this is not cleaned to `false` again? I know this is a workaround but I'm not sure if we'd better use a specific temporary property for this instead of `HasMasterPassword` that is used in other places. What do you think @mpbw2 ?

Reply:
@mpbw2 agreed let's move `HasMasterPassword` checks to _userVerificationService

Reply:
Actually _userVerificationService.HasMasterPasswordAsync() does a couple more checks, it is replacing the previous check for keyconnector usage. 
I'm not 100% that i can use `_userVerificationService.HasMasterPasswordAsync()` instead of the direct  `_stateService.GetAccountDecryptionOptions().HasMasterPassword`.

Reply:
```
public async Task<bool> HasMasterPasswordAsync(bool checkMasterKeyHash = false)
{
    async Task<bool> CheckMasterKeyHashAsync()
    {
        return !checkMasterKeyHash || await _cryptoService.GetMasterKeyHashAsync() != null;
    };

    var decryptOptions = await _stateService.GetAccountDecryptionOptions();
    if (decryptOptions != null)
    {
        return decryptOptions.HasMasterPassword && await CheckMasterKeyHashAsync();
    }

    return !await _keyConnectorService.GetUsesKeyConnectorAsync() && await CheckMasterKeyHashAsync();
}
```

Reply:
Can *TDE* use *KeyConnector*? If not, I don't see a reason why you can't use `HasMasterPasswordAsync(false)`. If possible I'd use the same one and try to centralize everything there with the different flows / scenarios so we avoid having different methods to check the "same" thing.

Reply:
After discussing this with the auth team, and the same problem happening in the other clients, we decided it was best to create a follow up tech dept ticket to address this code fragmentation problem. 
[Ticket PM-4579](https://bitwarden.atlassian.net/browse/PM-4579)